website/docs/setup/guides/unit-testing.md

```diff
@@ -0,0 +1,245 @@
+
+```toml
+[[tests]]
+  name = "check_simple_log"
+
+  [tests.input]
+    insert_at = "foo"
+    type = "raw"
+    value = "2019-11-28T12:00:00+00:00 info Sorry, I'm busy this week Cecil"
+```
+
+Here we've specified that our test should begin by injecting an event at the
+transform `foo`. The `raw` input type creates a log with only a `message` field
+and `timestamp` (set to the time of the test), where `message` is populated with
+the contents of the `value` field.
+
+## Conditions
+
+This test won't work in its current state because there's nothing to check. In
+order to perform checks with this unit test we define an expected output:
+
+```toml
+[[tests]]
+  name = "check_simple_log"
+
+  [tests.input]
+    insert_at = "foo"
+    type = "raw"
+    value = "2019-11-28T12:00:00+00:00 info Sorry, I'm busy this week Cecil"
+
+  [[tests.outputs]]
+    extract_from = "baz"
+
+    [[tests.outputs.conditions]]
+      type = "check_fields"
+      "level.exists" = false
+      "new_field.equals" = "this is a static value"
+      "timestamp.equals" = "2019-11-28T12:00:00+00:00"
+      "message.equals" = "Sorry, I'm busy this week Cecil"
+```
+
+You can define any number of expected outputs, where we must specify at which
+transform the output events should be extracted for checking. This allows you to
+check the events from different transforms in a single test. For our purposes we
+only need to check the output of `baz`.
+
+An output can also have any number of conditions to check. In order for the test
+to pass each condition for an output must resolve to `true`. It's possible for a
+topology to result in >1 events extracted from a single transform, in which case
+a condition must pass for one or more of the extracted events in order for the
+test to pass.
+
+The only condition we've defined here is a `check_fields` type. This is
+currently the _only_ condition type on offer, and it allows us to specify any
+number of field queries (of the format `"<field>.<predicate>" = "<argument>"`).
+
+## Executing
+
+With this test appended to the bottom of our config we are now able to execute
+it. Executing tests within a config file can be done with the `test` subcommand:
+
+```bash
+vector test ./example.toml
+```
+
+Doing this results in the following output:
+
+```sh
+$ vector test ./example.toml 
+Running ./example.toml tests
+Test ./example.toml: check_simple_log ... failed
+
+failures:
+
+--- ./example.toml ---
+
+Test 'check_simple_log':
+check transform 'baz' failed conditions: [ 0 ], payloads (encoded in JSON format):
+  {"timestamp":"2019-11-28T12:00:00+00:00","message":"Sorry, I'm busy this week Cecil"}
+```
+
+Woops! Something isn't right. Unfortunately we're only told which
+output-condition failed, not which predicate of our `check_fields` condition
+specifically caused the failure. If we refactor our test slightly we can make it
+clearer by breaking our condition down to one per predicate:
```

Comment:
Is there any way we could improve this so a user wouldn't have to break out the conditions? Or, open an issue to do so.

Reply:
It would mean fleshing out the condition API so that it's possible to return failure context ("field `foo` predicate `exists` failed as the field does not exist"). I think this is one item of a few we should discuss when reviewing conditions and their place in the vector universe.